contracts/libraries/PositionValue.sol

```diff
@@ -0,0 +1,226 @@
+                    tokensOwed0: tokensOwed0,
+                    tokensOwed1: tokensOwed1
+                })
+            );
+    }
+
+    function _fees(INonfungiblePositionManager positionManager, FeeParams memory feeParams)
+        private
+        view
+        returns (uint256 amount0, uint256 amount1)
+    {
+        (uint256 poolFeeGrowthInside0LastX128, uint256 poolFeeGrowthInside1LastX128) =
+            _getFeeGrowthInside(
+                positionManager,
+                feeParams.token0,
+                feeParams.token1,
+                feeParams.fee,
+                feeParams.tickLower,
+                feeParams.tickUpper
+            );
+
+        amount0 =
+            FullMath.mulDiv(
+                poolFeeGrowthInside0LastX128 - feeParams.positionFeeGrowthInside0LastX128,
+                feeParams.liquidity,
+                FixedPoint128.Q128
+            ) +
+            feeParams.tokensOwed0;
+
+        amount1 =
+            FullMath.mulDiv(
+                poolFeeGrowthInside1LastX128 - feeParams.positionFeeGrowthInside1LastX128,
+                feeParams.liquidity,
+                FixedPoint128.Q128
+            ) +
+            feeParams.tokensOwed1;
+    }
+
+    function _getFeeGrowthInside(
+        INonfungiblePositionManager positionManager,
+        address token0,
+        address token1,
+        uint24 fee,
+        int24 tickLower,
+        int24 tickUpper
+    ) private view returns (uint256 feeGrowthInside0X128, uint256 feeGrowthInside1X128) {
+        IUniswapV3Pool pool =
+            IUniswapV3Pool(
+                PoolAddress.computeAddress(
+                    positionManager.factory(),
+                    PoolAddress.PoolKey({token0: token0, token1: token1, fee: fee})
+                )
+            );
+
+        (, int24 tickCurrent, , , , , ) = pool.slot0();
+        (, , uint256 lowerFeeGrowthOutside0X128, uint256 lowerFeeGrowthOutside1X128, , , , ) = pool.ticks(tickLower);
+        (, , uint256 upperFeeGrowthOutside0X128, uint256 upperFeeGrowthOutside1X128, , , , ) = pool.ticks(tickUpper);
+
+        if (tickCurrent < tickLower) {
+            feeGrowthInside0X128 = lowerFeeGrowthOutside0X128 - upperFeeGrowthOutside0X128;
+            feeGrowthInside1X128 = lowerFeeGrowthOutside1X128 - upperFeeGrowthOutside1X128;
+        } else if (tickCurrent < tickUpper) {
+            uint256 feeGrowthGlobal0X128 = pool.feeGrowthGlobal0X128();
+            uint256 feeGrowthGlobal1X128 = pool.feeGrowthGlobal1X128();
+            feeGrowthInside0X128 = feeGrowthGlobal0X128 - lowerFeeGrowthOutside0X128 - upperFeeGrowthOutside0X128;
+            feeGrowthInside1X128 = feeGrowthGlobal1X128 - lowerFeeGrowthOutside1X128 - upperFeeGrowthOutside1X128;
+        } else {
+            feeGrowthInside0X128 = upperFeeGrowthOutside0X128 - lowerFeeGrowthOutside0X128;
+            feeGrowthInside1X128 = upperFeeGrowthOutside1X128 - lowerFeeGrowthOutside1X128;
+        }
+    }
+
+    function _constructParams(
```

Comment:
somewhat prefer just taking a gas hit and making this more readable since it's for public consumption, by just having fees and principal be public methods and having total just do `fees(nft, tokenId) + principal(nft,tokenId,sqrtRatioX96)`

if you can flatten the code out a bit as a result and remove the passing around of large structs, it might be a negligible difference